### PR TITLE
Weather

### PR DESCRIPTION
**Pull request by Garner071**
Raziel

### CODE REVIEW

File: plugins/weather.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+class WeatherPlugin(ircbot.plugin.Plugin):
+	@ircbot.plugin.command('weather')
+	def weather(self, cmd):
+		if len(cmd.args) < 2:
+			return 'Usage: !weather city state/country'
+		try:
+			url = 'http://api.wunderground.com/api/ba6f8fecb3824c22/geolookup/conditions/q/' + cmd.args[1] + '/' + cmd.args[0] + '.json'
+			response = urllib.request.urlopen(url, timeout=2).read()
+			json_string = json.loads(response.decode('utf-8'))
+		except (urllib.error.URLError, urllib.error.HTTPError) as e:
```
Comment (anlutro) [inline plugins/weather.py]: HTTPError is a subclass of URLError, so you don't need to catch that one. However, you should catch `socket.TimeoutError`.

File: plugins/weather.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+class WeatherPlugin(ircbot.plugin.Plugin):
+	@ircbot.plugin.command('weather')
+	def weather(self, cmd):
+		if len(cmd.args) < 2:
+			return 'Usage: !weather city state/country'
+		try:
+			url = 'http://api.wunderground.com/api/ba6f8fecb3824c22/geolookup/conditions/q/' + cmd.args[1] + '/' + cmd.args[0] + '.json'
+			response = urllib.request.urlopen(url, timeout=2).read()
+			json_string = json.loads(response.decode('utf-8'))
```
Comment (anlutro) [inline plugins/weather.py]: `data` is probably a better variable name - `json.loads` returns a dictionary object, not a string.

File: plugins/weather.py (hunk 1 above)
Comment (Garner071) [inline plugins/weather.py]: Ok will do that, thanks!

File: plugins/weather.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+		except (urllib.error.URLError, urllib.error.HTTPError) as e:
+			return 'An error occured, try again later!'
+		try:
+			city = json_string['location']['city']
+			state = json_string['location']['state']
+			weather = json_string['current_observation']['weather']
+			temperature_string = json_string['current_observation']['temperature_string']
+		except KeyError as e:
+			return 'Usage: !weather city state/country'
+		return 'Weather in ' + city + ': ' + weather.lower() + ', the temperature is ' + temperature_string
```
Comment (anlutro) [inline plugins/weather.py]: Try using str.format (interpolation) here instead of concatenation.

File: plugins/weather.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+			json_string = json.loads(response.decode('utf-8'))
+		except (urllib.error.URLError, urllib.error.HTTPError) as e:
+			return 'An error occured, try again later!'
+		try:
+			city = json_string['location']['city']
+			state = json_string['location']['state']
+			weather = json_string['current_observation']['weather']
+			temperature_string = json_string['current_observation']['temperature_string']
+		except KeyError as e:
+			return 'Usage: !weather city state/country'
```
Comment (anlutro) [inline plugins/weather.py]: If this exception is raised, that would mean the JSON returned from the webservice is invalid - so this error message could probably be changed to "Invalid data returned" or something like that.

File: plugins/weather.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	@ircbot.plugin.command('weather')
+	def weather(self, cmd):
+		if len(cmd.args) < 2:
+			return 'Usage: !weather city state/country'
+		try:
+			url = 'http://api.wunderground.com/api/ba6f8fecb3824c22/geolookup/conditions/q/' + cmd.args[1] + '/' + cmd.args[0] + '.json'
+			response = urllib.request.urlopen(url, timeout=2).read()
+			json_string = json.loads(response.decode('utf-8'))
+		except (urllib.error.URLError, urllib.error.HTTPError) as e:
+			return 'An error occured, try again later!'
```
Comment (anlutro) [inline plugins/weather.py]: Replace "An error" with "HTTP error" so that it's clear what's happening.

File: plugins/weather.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import json
+import ircbot.plugin
+
+class WeatherPlugin(ircbot.plugin.Plugin):
+	@ircbot.plugin.command('weather')
+	def weather(self, cmd):
+		if len(cmd.args) < 2:
+			return 'Usage: !weather city state/country'
+		try:
+			url = 'http://api.wunderground.com/api/ba6f8fecb3824c22/geolookup/conditions/q/' + cmd.args[1] + '/' + cmd.args[0] + '.json'
```
Comment (anlutro) [inline plugins/weather.py]: Change this to use str.format as well. Also see comment about how to get the wunderground API key from config.

**Comment by anlutro** [comment]
I'd add this:

```python
api_key = self.bot.config.get('wunderground_api_key')
if not api_key:
    return

url = 'http://...
```

That way, the api key is fetched from config.yml, and if no api key is set, nothing is done.

**Comment by anlutro** [comment]
Good enough! Thanks!

**Pull request merged**